### PR TITLE
feat(python): add MCPServer mount support for external connectors

### LINKED ISSUE #951: MCPServer.mount() for composing servers with external connectors
## Summary

Enable MCPServer to mount external connectors, allowing servers to compose local tools with remote tools from other MCP servers.

## Usage

```python
from mcp_use import MCPServer, StdioConnector, HttpConnector

server = MCPServer(name="composed-tools")

# Define local tools
@server.tool()
def my_local_tool(query: str) -> str:
    return f"Local result for {query}"

# Mount external servers with prefixes
playwright = StdioConnector(name="playwright", command="npx", args=["@playwright/mcp"])
filesystem = StdioConnector(name="filesystem", command="npx", args=["@modelcontextprotocol/server-filesystem"])

server.mount(playwright, prefix="browser")
server.mount(filesystem, prefix="fs")

# Server now exposes:
# - my_local_tool (local)
# - browser_click, browser_navigate (from playwright)
# - fs_read_file, fs_write_file (from filesystem)

server.run(transport="http", port=8080)
```

## Tool Composition Pattern

Create higher-level tools that orchestrate multiple MCP servers:

```python
@server.tool()
async def screenshot_and_save(url: str, filepath: str) -> str:
    """Navigate to URL, take screenshot, and save to file."""
    await server.call_tool("browser_navigate", {"url": url})
    screenshot = await server.call_tool("browser_screenshot", {})
    await server.call_tool("fs_write_file", {
        "path": filepath,
        "content": screenshot["data"],
    })
    return f"Screenshot saved to {filepath}"
```

## Implementation

```python
class MCPServer:
    def mount(self, connector: Connector, prefix: str | None = None) -> None:
        """Mount an external connector's tools on this server."""
        if prefix:
            connector = connector.with_prefix(prefix)
        self._mounted_connectors.append(connector)
```

## Acceptance Criteria
- [ ] Server.mount() accepts any Connector
- [ ] Mounted tools exposed via list_tools
- [ ] Tool calls routed to correct connector
- [ ] Prefixes applied correctly
- [ ] Server can call mounted tools internally

### PR DESCRIPTION
**Pull request by Sohamkayal4103**
# Pull Request Description

Adds `MCPServer.mount()` support for composing local server tools with tools from external MCP connectors. Mounted tools are exposed with prefixes through `tools/list`, routed through `tools/call`, and available for internal orchestration through `server.call_tool(...)`.

## Language / Project Scope

Check all that apply:
- [ ] TypeScript
- [x] Python
- [ ] Documentation only
- [ ] CI/CD or tooling

## Changes

Adds `MCPServer.mount()` support for composing local server tools with remote tools from external MCP connectors.

## Implementation Details

1. Added mounted connector registration to `MCPServer` with required prefix validation and duplicate-prefix checks.
2. Added mounted connector lifecycle handling so mounted connectors connect and initialize at server startup, build a cached prefixed tool map/tool list, and disconnect on shutdown.
3. Added a cached mounted tool map for routing prefixed tool names to `(connector, original_tool_name)` and a mounted tool cache for `tools/list` responses.
4. Used `model_copy(update={"name": prefixed_name})` when building mounted tool metadata so the server can expose prefixed tool names without mutating the original upstream `Tool` objects returned by the connector.
5. Wrapped `ListToolsRequest` handling to merge native server tools with mounted connector tools.
6. Wrapped `CallToolRequest` handling to resolve prefixed mounted tool names back to the correct connector and original upstream tool name.
7. Added `server.call_tool(...)` support for mounted tools only, matching the issue’s composition example.
8. Added a shared runner helper to wrap stdio and HTTP server execution with mounted connector startup/shutdown logic, avoiding duplicate lifecycle code across transports.
9. Added unit coverage for mount registration, lifecycle behavior, tool listing, tool routing, mounted internal calls, cache behavior, and runner cleanup.
10. No dependencies were added or changed.

---

## TypeScript Checklist

> Complete this section if your PR includes TypeScript changes

### Packages Modified

Check all packages that were modified:
- [ ] `docs`
- [ ] `tests`
- [ ] `cli`
- [ ] `create-mcp-use-app`
- [ ] `mcp-use` (server)
- [ ] `mcp-use` (client)
- [ ] `inspector`

### Pre-commit Checklist

Ensure all of the following have been completed:
- [ ] Ran `pnpm lint:fix` to auto-fix linting issues
- [ ] Ran `pnpm format` to format code with Prettier
- [ ] Ran `pnpm build` and build succeeds without errors
- [ ] Ran `pnpm changeset` to create a changeset (if this PR includes user-facing changes)
- [ ] Added or updated tests if needed
- [ ] Updated documentation in `docs/` folder if needed

---

## Python Checklist

> Complete this section if your PR includes Python changes

### Pre-commit Checklist

Ensure all of the following have been completed:
- [x] Code formatted with `ruff format`
- [x] Linting passes with `ruff check`
- [x] Added or updated tests if needed
- [ ] Updated documentation if needed

---

## Example Usage (Before)

```python
from mcp_use import MCPServer, StdioConnector

server = MCPServer(name="composed-tools")

@server.tool()
def my_local_tool(query: str) -> str:
    return f"Local result for {query}"

playwright = StdioConnector(name="playwright", command="npx", args=["@playwright/mcp"])
filesystem = StdioConnector(
    name="filesystem",
    command="npx",
    args=["@modelcontextprotocol/server-filesystem"],
)

# No generic server-side mounting support.
# External MCP tools had to be managed directly through connector objects.
await playwright.connect()
await playwright.initialize()
await filesystem.connect()
await filesystem.initialize()

await playwright.call_tool("navigate", {"url": "https://example.com"})
await filesystem.call_tool("write_file", {"path": "/tmp/out.txt", "content": "hello"})
```

## Example Usage (After)

```python
from mcp_use import MCPServer, StdioConnector, HttpConnector

server = MCPServer(name="composed-tools")

@server.tool()
def my_local_tool(query: str) -> str:
    return f"Local result for {query}"

playwright = StdioConnector(name="playwright", command="npx", args=["@playwright/mcp"])
filesystem = StdioConnector(
    name="filesystem",
    command="npx",
    args=["@modelcontextprotocol/server-filesystem"],
)

server.mount(playwright, prefix="browser")
server.mount(filesystem, prefix="fs")

@server.tool()
async def screenshot_and_save(url: str, filepath: str) -> str:
    await server.call_tool("browser_navigate", {"url": url})
    screenshot = await server.call_tool("browser_screenshot", {})
    await server.call_tool("fs_write_file", {"path": filepath, "content": screenshot["data"]})
    return f"Screenshot saved to {filepath}"

server.run(transport="http", port=8080)
```

## Documentation Updates

* No documentation files were updated.
* This change is covered by new unit tests.

## Testing

Describe how you tested these changes:
- Added `tests/unit/server/test_mount.py`
- Ran `pytest -vv tests/unit/server/test_mount.py`
- Ran `pytest -vv tests/unit/server`
- Ran `ruff check .`
- Ran `ruff format --check .`

Verified:
- mount registration
- empty/duplicate prefix validation
- mounted connector startup/shutdown lifecycle
- mounted tool exposure through `tools/list`
- mounted tool routing through `tools/call`
- mounted internal orchestration through `server.call_tool(...)`
- no repeated mounted tool cache rebuild for unknown mounted names after initialization

## Backwards Compatibility

These changes are backwards compatible.
- Existing local `MCPServer` tool behavior is unchanged.
- Mounted connector support is opt-in through `server.mount(...).`
- Existing connector APIs are unchanged.

## Related Issues

Closes #951

### CODE REVIEW

File: libraries/python/mcp_use/server/server.py [hunk 1]
```diff
@@ -318,6 +328,73 @@ def add(a: int, b: int) -> int:
...
+        """Mount an external connector's tools on this server."""
+        normalized_prefix = prefix.strip()
+        if not normalized_prefix:
+            raise ValueError("Mounted connector prefix must be non-empty")
+
+        if any(existing_prefix == normalized_prefix for existing_prefix, _ in self._mounted_connectors):
+            raise ValueError(f"Mounted connector prefix '{normalized_prefix}' is already registered")
+
+        self._mounted_connectors.append((normalized_prefix, connector))
+        self._mounted_tool_cache_built = False
```
[the remark this answers is not in the document]
Comment (Sohamkayal4103) [reply libraries/python/mcp_use/server/server.py:371]: Addressed in the latest push. `mount()` now normalizes prefixes using the same tool-name sanitization rule already used in the OpenAI / Anthropic / Google adapters, so spaces / hyphens / other invalid characters get converted to a provider-compatible prefix before registration. I also added test coverage for sanitization, truncation to 64 chars, empty-after-sanitization rejection, and duplicate detection after normalization.

File: libraries/python/mcp_use/server/server.py [hunk 2]
```diff
@@ -318,6 +328,73 @@ def add(a: int, b: int) -> int:
...
+
+        self._mounted_tool_map = mounted_tool_map
+        self._mounted_tools_cache = mounted_tools_cache
+        self._mounted_tool_cache_built = True
+
+    async def _ensure_mounted_tool_cache(self) -> None:
+        """Build mounted tool metadata once before serving lookups."""
+        if self._mounted_connectors and not self._mounted_tool_cache_built:
+            await self._build_mounted_tool_cache()
+
```
[the remark this answers is not in the document]
Comment (Sohamkayal4103) [reply libraries/python/mcp_use/server/server.py:420]: Addressed in the latest push. I added an `asyncio.Lock` around `_ensure_mounted_tool_cache()` and used a double-check pattern so concurrent requests don’t trigger duplicate `_build_mounted_tool_cache()` runs after the same invalidation. I also added a regression test that calls `_ensure_mounted_tool_cache()` concurrently and verifies the cache build only happens once.

File: libraries/python/mcp_use/server/server.py [hunk 3]
```diff
@@ -490,6 +570,51 @@ async def call_original(_: ServerMiddlewareContext[Any]) -> Any:
...
+                if not self._mounted_tools_cache:
+                    return result
+
+                tools_result = getattr(result, "root", result)
+                if not isinstance(tools_result, ListToolsResult):
+                    return result
+
+                merged_tools = [*tools_result.tools, *self._mounted_tools_cache]
+                merged_result = tools_result.model_copy(update={"tools": merged_tools})
```
[the remark this answers is not in the document]
Comment (Sohamkayal4103) [reply libraries/python/mcp_use/server/server.py:647]: Addressed in the latest push. `wrapped_list_tools()` now calls `_ensure_mounted_tool_cache()` before reading `_mounted_tools_cache`, and `mount()` now fully invalidates the mounted cache by clearing `_mounted_tool_map` / `_mounted_tools_cache` before marking it stale. I also added regression coverage for the “mount after initial cache build” case to verify `tools/list` rebuilds and returns the updated mounted tool set.

**Comment by Sohamkayal4103** [comment]
Hi @pietrozullo , Can you please review this PR?
